lecture-notes/3-bld.md

```diff
@@ -418,20 +420,284 @@ order:
+
 ### Implicit rules
+
+If you have ever had `foo.c` sitting around and no Makefile yet, but
+accidentally run `make`, you may have noticed Make... still do something? This
+is because of *implicit rules*. These "implicit rules" are automatically
+present any time you run `make` even if you don't have a Makefile at all!
+
+Make already knows how to compile C, C++, Fortran, and many other languages:
+inside Make are built-in pattern rules for common file extensions (e.g. `.c`,
+`.cpp`, `.o`).
+
+Because they are built-in, they support limited customization. If the implicit
+rule makes use of variables (such as `CC`, `CFLAGS`, and `LDFLAGS`), then you
+can customize execution by defining those variables.
+
+```console
+$ cat Makefile
+CFLAGS := -g
+hello: hello.o world.o
+$ make hello
+cc -g -c -o hello.o hello.c  # from implicit rule for compiling C
+cc -g -c -o world.o world.c  # from implicit rule for compiling C
+cc hello.o world.o -o hello  # from implicit rule for linking C
+rm -f hello.o  # by default, Make doesn't keep intermediate
+               # files between two pattern rules
+rm -f world.o
+$
+```
+
+Dependencies can be altered by writing a rule definition with no recipe. For
+example, the implicit `.c` to `.o` rule only has the `.c` file as a dependency:
+
+```make
+%.o: %.c
+	$(CC) $(CPPFLAGS) $(CFLAGS) -c
+```
+
+If you have any header files that are dependencies for a given object file, you
+probably want them included. To include the dependencies you want without
+rewriting the whole thing, skip the recipe portion of the rule; the implicit
+recipe will be used.
+
+```make
+hello.o: hello.c world.h
+	# still going to do $(CC) etc
+```
+
+Note that if you are, for example, adding more `.o` dependencies to a big C
+binary and still planning on using the implicit rule, you must make sure that
+the target matches at least one of the `.o` dependencies or the rule won't
+match:
+
+```make
+hello: hello.o world.o foo.o bar.o
+```
+
+Because `hello` shares a stem with `hello.o`, the rule will match. Also note
+that all of the other `.o` files are not the target of any rule! They use the
+implicit `%.o: %.c` pattern rule.
+
+#### Pitfalls
+
+It's not all fun and rainbows. Implicit rules are convenient, and you should
+use them if you want, but remember the following:
+
+* Header files not automatic prerequisites (but you can add them yourself, as
+  above)
+* Implicit rules are not explicit, which some people and projects value
+
+So they may not be the best for you and your team.
+
 ### Silencing commands
-### ?= and += assignment
+
+Build output can be noisy. When you run a command anywhere in a Makefile, Make
+prints the command (not its output---the command itself) before running it.
+This can be useful for debugging, but sometimes the command's only purpose is
+to print something else (e.g. `echo`). In this case, it's not helpful. To
+prevent the command from being printed, preface its line in your recipe with
+`@`.
```

Comment:
Might be a good place to also mention the `-` prefix, which makes a failure of that command non-fatal.